src/generic/grid.cpp

```diff
@@ -806,74 +806,181 @@ wxGridCellAttr *wxGridCellAttrData::GetAttr(int row, int col) const
     return attr;
 }
 
-void wxGridCellAttrData::UpdateAttrRows( size_t pos, int numRows )
+namespace
 {
-    size_t count = m_attrs.GetCount();
+
+void UpdateCellAttrRowsOrCols(wxGridCellWithAttrArray& attrs, int editPos,
+                              int editRowCount, int editColCount)
+{
+    wxASSERT( !editRowCount || !editColCount );
+
+    const bool isEditingRows = (editRowCount != 0);
+    const int editCount = (isEditingRows ? editRowCount : editColCount);
+
+    size_t count = attrs.GetCount();
     for ( size_t n = 0; n < count; n++ )
     {
-        wxGridCellCoords& coords = m_attrs[n].coords;
-        wxCoord row = coords.GetRow();
-        if ((size_t)row >= pos)
+        wxGridCellAttr* cellAttr = attrs[n].attr;
+        int cellRows, cellCols;
+        cellAttr->GetSize(&cellRows, &cellCols);
+
+        wxGridCellCoords& coords = attrs[n].coords;
+        const wxCoord cellRow = coords.GetRow(),
+                      cellCol = coords.GetCol(),
+                      cellPos = (isEditingRows ? cellRow : cellCol);
+
+        if ( cellPos < editPos )
         {
-            if (numRows > 0)
+            // This cell's coords aren't influenced by the editing, however
+            // do adjust a multicell's main size, if needed.
+            if ( GetCellSpan(cellRows, cellCols) == wxGrid::CellSpan_Main )
             {
-                // If rows inserted, increment row counter where necessary
-                coords.SetRow(row + numRows);
-            }
-            else if (numRows < 0)
-            {
-                // If rows deleted ...
-                if ((size_t)row >= pos - numRows)
+                int mainSize = isEditingRows ? cellRows : cellCols;
+                if ( cellPos + mainSize > editPos )
                 {
-                    // ...either decrement row counter (if row still exists)...
-                    coords.SetRow(row + numRows);
-                }
-                else
-                {
-                    // ...or remove the attribute
-                    m_attrs.RemoveAt(n);
-                    n--;
-                    count--;
+                    // Multicell is within affected range:
+                    // Adjust its size.
+
+                    if ( editCount >= 0 )
+                        mainSize += editCount;
+                    else
```

Comment:
I didn't leave such comments elsewhere, but we really need braces around the branches of this if statement just because the else branch is so, so long (the guideline in wx is to use braces for anything longer than one line).

Reply:
I wasn't aware that included counting commented lines, particularly if it would have been just one (I'll add them there as well, and not change the comment and move it in front of the `if` or after the consequent if it fits). But yes, definitely already needed here and then apparently also around the `if` (which I wasn't sure of, thanks). I prefer always using braces anyway, having been bitten once by commenting out only the consequent line.